utils/docker_utils.go

```diff
@@ -0,0 +1,153 @@
+		Image: config.Image,
+		Tty:   true,
+	}
+
+	b, err := cli.ContainerCreate(ctx, &dconfig, &hostConfig, nil, config.Name)
+
+	return b.ID, err
+}
+
+func StartContainer(cli *client.Client, id string) error {
+	log.Printf("Starting Container [%s]", id)
+	return cli.ContainerStart(context.Background(), id, types.ContainerStartOptions{})
+}
+
+func StopContainer(cli *client.Client, id string) error {
+	log.Printf("Stopping Container [%s]", id)
+	var timeout = 5 * time.Second
+	return cli.ContainerStop(context.Background(), id, &timeout)
+}
+
+func RemoveContainer(cli *client.Client, id string) error {
+	log.Printf("Removing Container [%s]", id)
+	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
+	defer cancel()
+	return cli.ContainerRemove(ctx, id, types.ContainerRemoveOptions{})
+}
+
+func Write(cli *client.Client, id string, content []string, dest string) error {
+	for _, c := range content {
+		c = strings.ReplaceAll(c, "'", "'\"'\"'")
+		cmd := fmt.Sprintf(`echo -E '%s' >> %s`, c, dest)
+		log.Printf("Writing [%s] on [%s] from Container [%s]", c, dest, id)
+		Exec(cli, id, cmd)
+	}
+	return nil
+}
+
+func Copy(cli *client.Client, id, src, dest string) error {
+	log.Printf("Copy [%s] to [%s] from Container [%s]", src, dest, id)
+	dat, _ := ioutil.ReadFile(src)
+	content := string(dat)
+	content = strings.ReplaceAll(content, "'", "'\"'\"'")
+	cmd := fmt.Sprintf("echo -E '%s' >| %s", content, dest)
+	return Exec(cli, id, cmd)
+}
+
+func Exec(cli *client.Client, id, cmd string) error {
+	log.Printf("Executing command [%s] on container [%s]", cmd, id)
+	config := types.ExecConfig{
+		Cmd: []string{"/bin/bash", "-c", cmd},
+	}
+	rid, _ := cli.ContainerExecCreate(context.Background(), id, config)
+	return cli.ContainerExecStart(context.Background(), rid.ID, types.ExecStartCheck{})
+}
+
+func Cat(cli *client.Client, id, path string) ([]byte, error) {
+	log.Printf("Getting content of file [%s]", path)
+	config := types.ExecConfig{
+		Tty:          true,
+		AttachStderr: true,
+		AttachStdout: true,
+		Cmd:          []string{"/bin/bash", "-c", "cat " + path},
+	}
+	rid, err := cli.ContainerExecCreate(context.Background(), id, config)
+	if err != nil {
+		log.Println(err)
+		log.Println("error on creating container exec")
+	}
+	log.Println(rid.ID)
+	hijack, _ := cli.ContainerExecAttach(context.Background(), rid.ID, config)
+	log.Println(hijack)
+	output := read(hijack.Conn)
+	log.Println(output)
+	return output, nil
+}
+
+func read(conn net.Conn) []byte {
+	result := make([]byte, 0)
+	b := make([]byte, 10)
+	for {
+		n, _ := conn.Read(b)
+		result = append(result, b...)
+		if n < len(b) {
+			break
+		}
+		b = make([]byte, 2*len(b))
+	}
+	return result
+}
+
+func Pull(cli *client.Client, image string) (io.ReadCloser, error) {
+	reader, err := cli.ImagePull(context.Background(), image, types.ImagePullOptions{})
+	return reader, err
+}
+
+func CheckImage(cli *client.Client, image string) (exist bool, err error) {
```

Comment:
what does this function check?

Reply:
it checks if the image is valid. It must exists on dockerhub.

Reply:
maybe the image registry can be customizable here

Reply:
Supporting registry other that dockerhub is a very good idea (not sure this  was the suggestion @emanueljoivo made, though)